pkg/registry/core/namespace/storage/storage_test.go

```diff
@@ -189,6 +191,143 @@ func TestDeleteNamespaceWithCompleteFinalizers(t *testing.T) {
 	}
 }
 
+func TestDeleteWithGCFinalizers(t *testing.T) {
+	storage, server := newStorage(t)
+	defer server.Terminate(t)
+	defer storage.store.DestroyFunc()
+
+	propagationBackground := metav1.DeletePropagationBackground
+	propagationForeground := metav1.DeletePropagationForeground
+	propagationOrphan := metav1.DeletePropagationOrphan
+
+	var tests = []struct {
+		name          string
+		deleteOptions *metav1.DeleteOptions
+
+		existingFinalizers  []string
+		remainingFinalizers map[string]bool
+	}{
+		{
+			name:          "propagation-nil",
+			deleteOptions: nil,
+			existingFinalizers: []string{
+				metav1.FinalizerOrphanDependents,
+				metav1.FinalizerDeleteDependents,
+			},
+			remainingFinalizers: map[string]bool{
+				metav1.FinalizerOrphanDependents: true,
+				metav1.FinalizerDeleteDependents: true,
+			},
+		},
+		{
+			name:                "propagation-nil-without-finalizers",
+			deleteOptions:       nil,
+			existingFinalizers:  []string{},
+			remainingFinalizers: map[string]bool{},
+		},
+		{
+			name: "propagation-background",
+			deleteOptions: &metav1.DeleteOptions{
+				PropagationPolicy: &propagationBackground,
+			},
+			existingFinalizers: []string{
+				metav1.FinalizerOrphanDependents,
+				metav1.FinalizerDeleteDependents,
+			},
+			remainingFinalizers: map[string]bool{},
+		},
+		{
+			name: "propagation-background-without-finalizers",
+			deleteOptions: &metav1.DeleteOptions{
+				PropagationPolicy: &propagationBackground,
+			},
+			existingFinalizers:  []string{},
+			remainingFinalizers: map[string]bool{},
+		},
+		{
+			name: "propagation-foreground",
+			deleteOptions: &metav1.DeleteOptions{
+				PropagationPolicy: &propagationForeground,
+			},
+			existingFinalizers: []string{
+				metav1.FinalizerOrphanDependents,
+				metav1.FinalizerDeleteDependents,
+			},
+			remainingFinalizers: map[string]bool{
+				metav1.FinalizerDeleteDependents: true,
+			},
+		},
+		{
+			name: "propagation-foreground-without-finalizers",
+			deleteOptions: &metav1.DeleteOptions{
+				PropagationPolicy: &propagationForeground,
+			},
+			existingFinalizers: []string{},
+			remainingFinalizers: map[string]bool{
+				metav1.FinalizerDeleteDependents: true,
+			},
+		},
+		{
+			name: "propagation-orphan",
+			deleteOptions: &metav1.DeleteOptions{
+				PropagationPolicy: &propagationOrphan,
+			},
+			existingFinalizers: []string{
+				metav1.FinalizerOrphanDependents,
+				metav1.FinalizerDeleteDependents,
+			},
+			remainingFinalizers: map[string]bool{
+				metav1.FinalizerOrphanDependents: true,
+			},
+		},
+		{
+			name: "propagation-orphan-without-finalizers",
+			deleteOptions: &metav1.DeleteOptions{
+				PropagationPolicy: &propagationOrphan,
+			},
+			existingFinalizers: []string{},
+			remainingFinalizers: map[string]bool{
+				metav1.FinalizerOrphanDependents: true,
+			},
+		},
```

Comment:
Can you add a case if deleteOptions.orphanDependents=true?

Also, could you add a case where existing finalizer is "orphan", and the deleteOptions requires "foreground"?

Reply:
Yeah, all added.